File: Day338/PathCrossing.java
class Solution {
    public boolean isPathCrossing(String path) {
         Set<Pair<Integer, Integer>> set = new HashSet<>();
        int x = 0, y = 0;
        set.add(new Pair<>(x, y));

        for (char c : path.toCharArray()) {
            if (c == 'N') {
                y += 1;
            } else if (c == 'S') {
                y -= 1;
            } else if (c == 'E') {
                x += 1;
            } else if (c == 'W') {
                x -= 1;
            }

            if (set.contains(new Pair<>(x, y))) {
                return true;
            }

            set.add(new Pair<>(x, y));
        }

        return false;
    }
}

File: Day338/PathCrossing.go
type Pair struct {
	X, Y int
}

// isPathCrossing checks if the given path forms a crossing.
func isPathCrossing(path string) bool {
	set := make(map[Pair]struct{})
	x, y := 0, 0
	set[Pair{X: x, Y: y}] = struct{}{}

	for _, c := range path {
		switch c {
		case 'N':
			y++
		case 'S':
			y--
		case 'E':
			x++
		case 'W':
			x--
		}

		if _, exists := set[Pair{X: x, Y: y}]; exists {
			return true
		}

		set[Pair{X: x, Y: y}] = struct{}{}
	}

	return false
}
